File: day03/src/day03.py
import functools

with open('resources/input.txt', 'r') as f:
    lines = [l.strip() for l in f.readlines()]
    width = len(lines[0])
    threshold = len(lines) / 2

    def most_common_bit(i):
        count = sum(int(word[i]) for word in lines)
        return 1 if count > threshold else 0

    gamma_rate = functools.reduce(lambda x, y: x | y, [most_common_bit(i) << (width - 1 - i) for i in range(width)])
    epsilon_rate = gamma_rate ^ ((1 << width) - 1) # python has no unary bitwise negation operator, therefore this
    print(f'Part 1: {gamma_rate * epsilon_rate}')

File: day03/src/day03.f90
program day03
  implicit none
  
  integer, parameter :: count = 1000
  integer, parameter :: width = 12
  integer, dimension(count) :: xs
  integer :: ios, i, n, sum, b, gamma_rate, epsilon_rate

  open(1, file='resources/input.txt')

  do n = 1, count
    read(1, '(B32)', iostat=ios) xs(n)
    if (ios /= 0) exit
  end do

  close(1)

  gamma_rate = 0

  do i = 0, width - 1
    sum = 0
    do n = 1, count
      sum = sum + iand(ishft(xs(n), -i), 1)
    end do
    if (sum >= (count / 2)) then
      b = 1
    else
      b = 0
    end if
    gamma_rate = ior(gamma_rate, ishft(b, i))
  end do

  epsilon_rate = ieor(gamma_rate, (ishft(1, width) - 1))

  print *, 'Part 1:', gamma_rate * epsilon_rate

end program day03
